### PR TITLE
feat: add type hints to modules 03, 04, and 05 (#1167)

### LINKED ISSUE #1167: Add type hints for all the class, functions in `tinytorch`
* Ensure type hints  are added for all classes and functions in the codebase. 
* Type hints help catch edge cases early and improve code readability and maintainability.
* Without them, issues are only discovered after failing test cases, which slows down debugging and learning.
* This will also benefit the students working on assignments, as type hints provide guidance and reduce ambiguity.

Additional context:
I am currently completing the assignments by removing the provided solutions using a script and solving them independently.

### PR DESCRIPTION
**Pull request by imuday984**
Resolves #1167 for modules 03 and 04.

Following the discussion with @harishb00 in my previous PR, I have left 01_tensor and 02_activations for them to push. I have applied the suggested __future__ imports and TypeAlias patterns to these new modules for cleaner code.

Note: I added these directly to the source files (src/) but haven't run the tito module complete build step locally to update the generated package files. Let me know if a maintainer can trigger the build.
thank you.

### CODE REVIEW

**Review by Kabir1618 — APPROVED** [review]

**Review by imuday984 — COMMENTED** [review]
alright then will work on other files too

File: tinytorch/src/03_layers/03_layers.py [hunk 1]
```diff
@@ -899,32 +922,32 @@ class Sequential:
         >>> params = model.parameters()  # All parameters from all layers
     """
 
-    def __init__(self, *layers):
+    def __init__(self, *layers: Layer) -> None:
```
Comment (harishb00) [inline tinytorch/src/03_layers/03_layers.py]: * The contract as per the typehints says that this method supports variable number of `Layer`. So, someone calls it can call like func(layer1, layer2, ...)
* If you look at the implementation, it says that input can also be like
    * func([layer1, layer2]), func((layer1, layer2))
 * We should express this fact in the contract. so that students get a chance to write logic for these edgecases.
 * Something like `def __init__(self, *layers: Union[Layer, list[Layer], tuple[Layer, ...]]) -> None:`
 * Please don't take it just like that. I may be wrong. Think through once if this makes sense. 
 * The intention behind the typehints are it's not just for IDEs but to help students to handle edgecases without us giving much details for every possible edgecase.
 * Suggestions apart, you are doing a good job for the community. Thank you so much. Keep up the good work!

**Review by codeCraft-Ritik — COMMENTED** [review]
Excellent work on this implementation — very clean and well-organized
